common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,225 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.io.filesystem;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Watches for modifications to the specified file and performs the specified action (in the form of a callback) upon
+ * modification detection.
+ *
+ * This implementation uses NIO and event-based Java WatchService to watch for any modifications to the specified file.
+ *
+ * Note that:
+ *
+ * - The specified actions may trigger slightly later than when the file is actually modified.
+ *
+ * - If there are multiple file modifications in quick succession, only the last one may trigger the specified action.
+ *
+ * - The callback twice under rare circumstances when a file is modified. The Java {@link WatchService} raises two
+ *   events per modification: one for the content and second for the modification time. While there is a de-duplication
+ *   logic in this class to ensure that only a single event is handled and notified to the callback, it may not work
+ *   at times. If the calling code needs to be notified exactly once at all times, it should consider using
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ * - This class won't work for files that is under a symbolic link (pointing to a directory). This is because the
+ *   Java {@link WatchService} does not raise events for modifications of such files. Consider using
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ */
+@Slf4j
+public class FileModificationEventWatcher extends Thread implements FileModificationMonitor {
```

Comment:
Since you have chosen to use an entire thread for this task, can you please give it an appropriate name so that when we do a thread dump we know which thread is which?

Reply:
The thread already has a name given to it here: https://github.com/pravega/pravega/blob/f96e9ccfdf2c4faeeb06d289cefdbeeb888c837e/common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java#L112